File: 141.环形链表/141.js
/**
 * 快慢指针解法
 * 时间复杂度：O(N)，其中 N 是链表中的节点数
 * 空间复杂度：O(1)
 */

/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

/**
 * @param {ListNode} head
 * @return {boolean}
 */
 var hasCycle = function(head) {
    if(!head){
        return false
    }
    let fast=head
    let slow=head
    
    while(fast && fast.next){
        fast=fast.next.next
        slow=slow.next
        if(fast==slow){
            return true
        }
    }
    return false
};

File: 141.环形链表/141.java
/**
 *  快慢指针解法
 */

/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public boolean hasCycle(ListNode head) {
        if(head==null || head.next==null){
            return false;
        }
        ListNode slow=head;
        ListNode fast=head.next;
        while(slow!=fast){
            //如果fast或者fast.next为空,则没有环，是一条尾指针为null的链表
            if(fast==null || fast.next==null){
                return false;
            }
            slow=slow.next;
            fast=fast.next.next;
        }
        return true;
    }
}